aspects/schema.go

```diff
@@ -60,8 +60,24 @@ func ParseSchema(raw []byte) (*StorageSchema, error) {
 		return nil, fmt.Errorf(`cannot parse top level schema: must have a "schema" constraint`)
 	}
 
-	// TODO: check "types" here and parse the user-defined types
-	schema := &StorageSchema{}
+	schema := new(StorageSchema)
+	if val, ok := schemaDef["types"]; ok {
+		var userTypes map[string]json.RawMessage
+		if err := json.Unmarshal(val, &userTypes); err != nil {
+			return nil, fmt.Errorf(`cannot parse user-defined types map: %w`, err)
+		}
+
+		schema.userTypes = make(map[string]parser, len(userTypes))
+		for userTypeName, typeDef := range userTypes {
```

Comment:
we should have a regexp for what are valid userTypeName, might be something to clarify in the spec

Reply:
Ah good point, do you have something in mind? Maybe 3-20 lowercase letters (or alphanumerics?) optionally separated with one or two dashes? Would support things like: "age", "snap-name", "acme-type"

Reply:
let's start with something like subkeyRegex/validSubkey and the discuss with the larger group?